### PR TITLE
CPS-420: Limit Query Result To Accessible Case Type Categories

### PR DESCRIPTION
**Pull request by shahrukh-compucorp**
## Overview
This pr limits the query result to only accessible case categories for current user if the query includes case entity. For example on advance search if we use case filters then only this limiting will occur. This pr will impact all queries that are ran by api, search or ui that includes case entity in a such a way that the result set is only restricted to the case categories that the current user has access to.

## Before
Previously this work was done in a `CRM/Civicase/BAO/Query/CaseCategory.php` hook but that was only for advance search.

## After
This pr removes that advance search specif work and introduces a more generic way which will work globally on every time a query is made that involves case entity.

## Technical Details
This pr utilizes the `hook_civicrm_selectWhereClause` hook and adds the clauses for the case entity as per the permissions of current user.

### CODE REVIEW

File: CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+use CRM_Civicase_Helper_CaseCategory as CaseCategoryHelper;
+
+/**
+ * Restrict search to accessible case categories if case entity is involved.
+ */
+class CRM_Civicase_Hook_SelectWhereClause_RestrictSearchToAccessibleCaseCategories {
```
Comment (tunbola) [inline CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php:8]: This name `RestrictSearchToAccessibleCaseCategories` is not appropriate again because this hook will do more than restrict search. It will affect more than search page and will affect selecting cases via api, ui, or searches. Everywhere the Case entity is being called will be affected. Maybe a name like `LimitCaseQueryToAccessibleCaseCategories` or something better.

File: CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      $clauses['case_type_id'] = 'IS NULL';
+    }
+    else {
+      $accessibleCaseCategoriesIds = implode(',', array_keys($accessibleCaseCategories));
+      $clauses['case_type_id'][] = "IN (SELECT id FROM civicrm_case_type WHERE case_type_category in ($accessibleCaseCategoriesIds))";
+    }
+  }
+
+  /**
+   * Whether the hook should run or not.
```
Comment (tunbola) [inline CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php:44]: Can we change this description to `Should Run when the entity is Case` or something more descriptive.

File: civicase.php [hunk 3]
```diff
@@ -573,6 +573,14 @@ function civicase_civicrm_selectWhereClause($entity, &$clauses) {
   if ($entity === 'Case') {
     unset($clauses['id']);
   }
```
Comment (tunbola) [inline civicase.php:575]: Can we move this logic to the new hook since it concerns the Case entity too? Or what do you think?

File: CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  public function run($entity, array &$clauses) {
+
+    if (!$this->shouldRun($entity)) {
+      return;
+    }
+    $accessibleCaseCategories = CaseCategoryHelper::getAccessibleCaseTypeCategories();
+    $params = (json_decode($_REQUEST['json'], TRUE));
+
+    if (
+      isset($params[0][2]['case_type_id.case_type_category']) &&
```
Comment (tunbola) [inline CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php:27]: There are so many uncertainties here. 
- How are we sure that this will always be the position of the `case_type_id.case_type_category ` param in the `params` array?
- The `params` array is gotten from the `$_REQUEST['json']`, will this always be populated when the Case API is called from PHP and from Frontend and also when invoked from searches?

File: CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      (
+        (in_array(ucfirst($params[0][2]['case_type_id.case_type_category']), $accessibleCaseCategories)) ||
+        in_array($params[0][2]['case_type_id.case_type_category'], array_keys($accessibleCaseCategories))
+      )) {
+      return;
+    }
+    elseif (isset($params[0][2]['case_type_id.case_type_category'])) {
+      $clauses['case_type_id'] = 'IS NULL';
```
Comment (tunbola) [inline CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php:35]: Can you add a comment inline why we need to do this?

**Comment by tunbola** [comment]
@shahrukh-compucorp , Can we add more details to the PR overview as regards how this PR will affect Case globally in search, UI, API calls and also change the PR topic as this change does not only affect the search. 
Also if this gets sent for testing we need to let the QA the magnitude of this change and list the key areas to test for regressions.

**Review by tunbola — CHANGES_REQUESTED** [review]

File: civicase.php (hunk 3 above)
Comment (shahrukh-compucorp) [reply civicase.php:575]: yes we should move it to a separate hook but I think we should do that in a separate ticket

File: CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategories.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+   *   Entity name.
+   * @param array $clauses
+   *   Where clauses.
+   */
+  public function run($entity, array &$clauses) {
+
+    if (!$this->shouldRun($entity)) {
+      return;
+    }
+    $accessibleCaseCategories = CaseCategoryHelper::getAccessibleCaseTypeCategories();
```
Comment (tunbola) [inline CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategories.php:23]: For administrators or users that have access to all case categories, we don't need to add the clause at all. We can use this function to avoid that: https://github.com/compucorp/uk.co.compucorp.civicase/blob/50aa75dde19e1828b13cb6614795809e105adcad/CRM/Civicase/BAO/Query/CaseCategory.php#L75-L80

**Review by tunbola — CHANGES_REQUESTED** [review]

File: tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    );
+  }
+
+  /**
+   * Test that no records are returned if user has no permissions.
+   */
+  public function testNoCasesAreReturnedForNoPermissions() {
+    $contact = ContactFabricator::fabricate();
+    $this->registerCurrentLoggedInContactInSession($contact['id']);
+    $this->setPermissions(['access my cases and activities']);
```
Comment (tunbola) [inline tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php:65]: We don't really need to set a permission here, the test says when no permission, you can just pass an empty array.

File: tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    $result = civicrm_api3('Case', 'get', [
+      'contact_id' => static::$client['id'],
+      'check_permissions' => 1,
```
Comment (tunbola) [inline tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php:69]: If we don't pass the check permission parameter, the hook won't be triggered?

File: tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    $contact = ContactFabricator::fabricate();
+    $this->registerCurrentLoggedInContactInSession($contact['id']);
+    $this->setPermissions(['basic case information']);
+
+    $result = civicrm_api3('Case', 'get', [
+      'contact_id' => static::$client['id'],
+      'check_permissions' => 1,
+    ]);
+
+    $this->assertEquals($result['count'], 1);
```
Comment (tunbola) [inline tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php:104]: Apart from checking that one case is returned, we need to check that the correct case is returned by asserting that it is the case that belongs to the case type with the case category.

**Review by tunbola — CHANGES_REQUESTED** [review]

File: tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+use CRM_Civicase_Test_Fabricator_Contact as ContactFabricator;
+use CRM_Civicase_Test_Fabricator_CaseType as CaseTypeFabricator;
+use CRM_Civicase_Test_Fabricator_Case as CaseFabricator;
+
+/**
+ * Tests for limiting case query to accessible case categories.
+ *
+ * @group headless
+ */
+class CRM_Civicase_Hook_SelectWhereClause_LimitCaseQueryToAccessibleCaseCategoriesTest extends BaseHeadlessTest {
```
Comment (tunbola) [inline tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php:12]: Since most of the API calls from the FE comes with the case_type.category parameter pre-filled. Can we have tests that reflect this also. E.g a call requesting for cases of category awards when the user does not have basic award information. Or a call requesting for the cases of category award when the user has both basic case and basic award information permission, will both cases be returned or just the one the user requested for?

File: tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php (hunk 7 above)
Comment (shahrukh-compucorp) [reply tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php:65]: Actually I need this permission to call for case api that is called in the same test below

File: tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php (hunk 8 above)
Comment (shahrukh-compucorp) [reply tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php:69]: yes

File: tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    $result = civicrm_api3('Case', 'getcaselist', [
+      'contact_id' => static::$client['id'],
+      'case_type_id.case_type_category' => [
+        'IN' => ['award'],
+      ],
+      'check_permissions' => 1,
+      'sequential' => 1,
+    ]);
+
+    $this->assertEquals($result['count'], 0);
```
Comment (tunbola) [inline tests/phpunit/CRM/Civicase/Hook/SelectWhereClause/LimitCaseQueryToAccessibleCaseCategoriesTest.php:137]: Not sure about the test title. The title says `testOnlyAccessibleCasesAreReturned` but no results are returned. Maybe we can rename to  `testNoCasesAreReturnedWhenUserTriesToFetchCasesForCategoryItDoesNotHaveAccessTo` or some better title.

Also we need one more test where the user is able to fetch for the case category he has access to and we ensure that only such cases are returned and it does not leak into other case type categories when the case type category parameter is sent, i.e `testCasesAreReturnedOnlyForCasesUserHasAccessToWhenCaseCategoryParameterIsSent`

**Review by tunbola — CHANGES_REQUESTED** [review]

**Review by tunbola — APPROVED** [review]
